[AST] Remove ParenType

hamishknight opened the pull request:
Today ParenType is used:

1. As the type of ParenExpr
2. As the payload type of an unlabeled single associated value enum case (and the type of ParenPattern).
3. As the type for an `(X)` TypeRepr

For 1, this leads to some odd behavior, e.g the type of `(5.0 * 5).squareRoot()` is `(Double)`. For 2, we should be checking the arity of the enum case constructor parameters and the presence of ParenPattern respectively. Eventually we ought to consider replacing Paren/TuplePattern with a PatternList node, similar to ArgumentList.

3 is one case where it could be argued that there's some utility in preserving the sugar of the type that the user wrote. However it's really not clear to me that this is particularly desirable since a bunch of diagnostic logic is already stripping ParenTypes. In cases where we care about how the type was written in source, we really ought to be consulting the TypeRepr.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/llvm-project/pull/4769

@swift-ci please test source compatibility

hamishknight commented on lib/APIDigester/ModuleAnalyzerNodes.cpp:
```diff
@@ -1662,12 +1659,6 @@ SwiftDeclCollector::constructTypeNode(Type T, TypeInitInfo Info) {
 
   SDKNode* Root = SDKNodeInitInfo(Ctx, T, Info).createSDKNode(SDKNodeKind::TypeNominal);
 
-  // Keep paren type as a stand-alone level.
-  if (auto *PT = dyn_cast<ParenType>(T.getPointer())) {
-    Root->addChild(constructTypeNode(PT->getSinglyDesugaredType()));
-    return Root;
-  }
-
```
@nkcsgexi Any concerns with removing ParenType from the api-digester? This should only affect when it runs in non-ABI mode

hamishknight commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -5881,14 +5881,17 @@ NeverNullType TypeResolver::resolveInverseType(InverseTypeRepr *repr,
   // metatype so we can look at the instance type. We'll re-wrap at the end.
-  ExistentialMetatypeType *existentialTy =
-      dyn_cast<ExistentialMetatypeType>(ty.get().getPointer());
+  // Note we don't look through parens.
+  ExistentialMetatypeType *existentialTy = nullptr;
+  if (!constraintRepr->isParenType())
+    existentialTy = dyn_cast<ExistentialMetatypeType>(ty.get().getPointer());
```
@kavon Is it intentional that this doesn't look through parens, e.g we accept `any ~Copyable.Type`, but not `any ~(Copyable.Type)`? For now I've just maintained the current behavior

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test source compatibility

hamishknight commented:
Split out a couple of PRs to hopefully make this a bit easier to land:

- https://github.com/swiftlang/swift/pull/77202
- https://github.com/swiftlang/swift/pull/77203

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test source compatibility

slavapestov reviewed (approved):
Can't wait to see this land!

hamishknight commented on include/swift/Demangling/TypeDecoder.h:
```diff
@@ -1470,16 +1470,6 @@ class TypeDecoder {
 
       return Builder.createDictionaryType(key.getType(), value.getType());
     }
-    case NodeKind::SugaredParen: {
```
Ah okay, does that mean we also need to keep the logic in Demangler.cpp?

hamishknight replied:
Updated with the restored demangling code

hamishknight committed:
[Sema] Avoid relying on ParenType in `coercePatternToType`

I missed this in my other Sema ParenType PR, avoid
relying on ParenType here.

hamishknight committed:
[Sema] Make explicit that `resolveInverseType` doesn't look through parens

Check for `isParenType` on the TypeRepr to avoid
relying on getting that behavior by casting the
type pointer.

hamishknight committed:
[AST] Remove ParenType

Today ParenType is used:

1. As the type of ParenExpr
2. As the payload type of an unlabeled single
   associated value enum case (and the type of
   ParenPattern).
3. As the type for an `(X)` TypeRepr

For 1, this leads to some odd behavior, e.g the
type of `(5.0 * 5).squareRoot()` is `(Double)`. For
2, we should be checking the arity of the enum case
constructor parameters and the presence of
ParenPattern respectively. Eventually we ought to
consider replacing Paren/TuplePattern with a
PatternList node, similar to ArgumentList.

3 is one case where it could be argued that there's
some utility in preserving the sugar of the type
that the user wrote. However it's really not clear
to me that this is particularly desirable since a
bunch of diagnostic logic is already stripping
ParenTypes. In cases where we care about how the
type was written in source, we really ought to be
consulting the TypeRepr.

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test Windows

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test

xedin reviewed (approved):
LGTM! It's slightly unfortunate that we loose the sugar sometimes in the diagnostics but I don't think it's critical.

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please smoke test macOS

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/4769

@swift-ci please test macOS

hamishknight commented:
🚢
